File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        hashmap = {}

        left, ans = 0, 0 
        for right in range(len(nums)):
            if nums[right] in hashmap:
                hashmap[nums[right]] += 1
            else:
                hashmap[nums[right]] = 1

            while hashmap[nums[right]] > k:
                hashmap[nums[left]] -= 1
                left += 1
            
            ans = max(ans, right-left+1)
        
        return ans

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        unordered_map<int, int> hashmap;

        int ans = 0;
        int n = nums.size(), left = 0;
        for(int right = 0; right < n; right++){
            hashmap[nums[right]]++;
            while(left <= right && hashmap[nums[right]] > k){
                hashmap[nums[left]]--;
                left++;
            }

            ans = max(ans, right-left+1);
        }

        return ans;
    }
};